### DISCUSSION TITLE
Getting error reason: Ambiguous path separator while fetching images from nested folders

### DISCUSSION THREAD

**Discussion by arunadlprivate** [cantaloupe-project/cantaloupe#651] [Q&A]
Getting error reason: Ambiguous path separator while fetching images from nested folders .If imaged moved to root of the prefix path it is working.But URLs including %2F not working and giving error. cantaloupe version 5.0.6

**Comment by dirkroorda** [comment]
I encounter it too, over https, but not over http:

In my set up, this works:

```
http://127.0.0.1:8182/iiif/3/03%2F03_0007_001-r.jpg/full/max/0/default.jpg
```

And this gives the `Bad message 400: reason: Ambiguous path separator` error:

```
https://127.0.0.1:8183/iiif/3/03%2F03_0007_001-r.jpg/full/max/0/default.jpg
```

**Reply by dirkroorda** [reply]
The setting is: directly running a docker image with

```
    docker run -d -p 8182:8182 -p 8183:8183 \
      -e "CANTALOUPE_ENDPOINT_ADMIN_SECRET=secret" \
      -e "CANTALOUPE_ENDPOINT_ADMIN_ENABLED=true" \
      -e "CANTALOUPE_HTTP_ENABLED=true" \
      -e "CANTALOUPE_HTTPS_ENABLED=true" \
      -e "CANTALOUPE_HTTPS_HOST=0.0.0.0" \
      -e "CANTALOUPE_HTTPS_PORT=8183" \
      -e "CANTALOUPE_HTTPS_KEY_STORE_TYPE=JKS" \
      -e "CANTALOUPE_HTTPS_KEY_STORE_PASSWORD=******" \
      -e "CANTALOUPE_HTTPS_KEY_STORE_PATH=/keys/keystore.jks" \
      -e "CANTALOUPE_HTTPS_KEY_PASSWORD=******" \
      --name project \
      -v /Users/me/backend/scans:/imageroot \
      -v /Users/me/backend/keys:/keys \
      uclalibrary/cantaloupe:5.0.6-4
```

**Reply by DiegoPino** [reply]
Hi 

I have the impression jetty (when dealing with internal https) does a different type of slash decoding/encoding and there are chances that without a proxy (e.g nginx which is what we use here) they are being not decoded/passed to the backend correctly. Maybe the http implementation is less opinionated.  Please check the documentation here https://cantaloupe-project.github.io/manual/5.0/sources.html#FilesystemSource and maybe try a different separator (at least as a test so we can be sure it is jetty) or try putting cantaloupe (always safe and also faster) behind a proxy that respects precoded slashes correctly, like nginx.

There is a whole issue in https://github.com/jetty/jetty.project/issues/7713 about encodings.

**Reply by glenrobson** [reply]
In the cantaloupe dev meeting and Kevin is going to have a look at the UCLA docker to see if this can be replicated.

**Reply by yls2git** [reply]
Hi All, 

In the ‘cantaloupe.property’ file, set a string as “slash_substitute” can be set. Maybe this will help with the encoding.

slash_substitute = __ (you can use two underscores)

Substitute %3A = : and %2F = / in the URL to access subfolders.

Best,

**Reply by janhoy** [reply]
Hi, I just upgraded to 5.0.6, and this started happening, with the exact same client code talking to cantaloupe. My identifier is a file path, encoded with `foo%2Fbar.txt`, and I'm using Java's `HttpClient` to initiate requests to the image server.

Found this jetty issue https://github.com/jetty/jetty.project/issues/7766 "Jetty rejects encoded URIs with HTTP/2", which lead me to try explicitly making the connection over HTTP/1.1 instead, and that solved it for me.

Perhaps declaring `HttpCompliance.LEGACY` in the [server config](https://github.com/cantaloupe-project/cantaloupe/blob/develop/src/main/java/edu/illinois/library/cantaloupe/ApplicationServer.java#L267) is a quick way to be more lenient?

**Reply by janhoy** [reply]
Btw, reading more code I found this line https://github.com/cantaloupe-project/cantaloupe/blob/develop/src/main/java/edu/illinois/library/cantaloupe/ApplicationServer.java#L303 which suggests that Cantaloupe prefers http1 when listening for https, which it does not do for http. Strange then that I get the problem for HTTP with HTTP/2 and not HTTP/1.1, but @dirkroorda gets the problem only with https.

Can you perhaps try to to force http/2 with cURL for plain http, and see if that triggers the same issue, e.g.?

```bash
curl --http2 -v http://localhost:8182/iiif/2/path%2Fmy.tif/full/max/0/default.jpg
```

**Reply by janhoy** [reply]
@dirkroorda Would you have a chance to test out HTTP1 vs HTTP2 so we can document a workaround for the issue in next release?

**Reply by janhoy** [reply]
I believe that this issue is first and foremost a Jetty 9.4.x issue. In Jetty 11+ there is a way to configure `HttpConfiguration.setHttpCompliance()` and `HttpConfiguration.setUriCompliance()` individually, and it the UriCompliance will default to a setting which is  a bit more lenient. I could not reproduce the issue from development branch.

So for all 5.0.6 users, forcing HTTP1.1 might be the best workaround. And for Cantaloupe 6 release, I believe the defaults are OK already, but let's test it before release so we can decide whether to add some configuration for `uri_compliance`...
